lib/dockerregistry/inventory.go

```diff
@@ -1171,6 +1236,143 @@ func (sc *SyncContext) ReadRegistries(
+				if registryContext.Name == registryName {
+					rc = registryContext
+				}
+			}
+			for imageName, digestTags := range rii {
+				for digest, tagSlice := range digestTags {
+					if sc.DigestMediaType[digest] == cr.DockerManifestList {
+						// Create the request.
+						var req stream.ExternalRequest
+						var tag Tag
+						if len(tagSlice) > 0 {
+							// It could be that this ManifestList has been
+							// tagged multiple times. Just grab the first tag.
+							tag = tagSlice[0]
+						}
+						gmlc := GManifestListContext{
+							RegistryContext: rc,
+							ImageName:       imageName,
+							Tag:             tag,
+							Digest:          digest}
+						req.RequestParams = gmlc
+						req.StreamProducer = mkProducer(sc, gmlc)
+						wg.Add(1)
+						reqs <- req
+					}
+				}
+			}
+		}
+	}
+
+	var processRequest ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+
+			// Now run the request (make network HTTP call with
+			// ExponentialBackoff()).
+			gManifestList, err := getGManifestListWrapper(req)
+			if err != nil {
+				// Skip this request if it has unrecoverable errors (even after
+				// ExponentialBackoff).
+				reqRes.Errors = Errors{
+					Error{
+						Context: "getGManifestListWrapper",
+						Error:   err}}
+				requestResults <- reqRes
+				wg.Add(-1)
+
+				continue
+			}
+
+			gmlc := req.RequestParams.(GManifestListContext)
+
+			for _, gManifest := range gManifestList.Manifests {
+				mutex.Lock()
+				sc.ParentDigest[gManifest.Digest] = gmlc.Digest
+				mutex.Unlock()
+			}
+
+			reqRes.Errors = Errors{}
+			requestResults <- reqRes
+
+			wg.Add(-1)
+		}
+	}
+	sc.ExecRequests(populateRequests, processRequest)
+}
+
+// FilterByTag removes all images in RegInvImage that do not match the
+// filterTag.
+func FilterByTag(rii RegInvImage, filterTag string) RegInvImage {
+	filtered := make(RegInvImage)
+	for imageName, digestTags := range rii {
+		for digest, tags := range digestTags {
+			for _, tag := range tags {
+				if string(tag) == filterTag {
+					if filtered[imageName] == nil {
+						filtered[imageName] = make(DigestTags)
+					}
+					filtered[imageName][digest] = append(
+						filtered[imageName][digest],
+						tag)
+				}
+			}
+		}
+	}
+	return filtered
+}
+
+// RemoveChildDigestEntries removes all tagless images in RegInvImage that are
+// referenced by ManifestLists in the Registries.
+func (sc *SyncContext) RemoveChildDigestEntries(rii RegInvImage) RegInvImage {
+	filtered := make(RegInvImage)
+	for imageName, digestTags := range rii {
+		for digest, tagSlice := range digestTags {
+			_, hasParent := sc.ParentDigest[digest]
+			// If this digest is referenced by a parent ManfestList, skip over
+			// it (do not add to "filtered") only if it is also tagless.
```

Comment:
You could simplify the language here e.g. "If this image digest is only referenced as part of a parent ManifestList (i.e. not directly tagged), we filter it out"

Reply:
Ack.